### DISCUSSION TITLE
descriptor set trouble

### DISCUSSION THREAD

**Discussion by beaumanvienna** [KhronosGroup/MoltenVK#2344] [General]
I get a seg fault in `vkQueueSubmit`:
[crash lucre.txt](https://github.com/user-attachments/files/17085523/crash.lucre.txt)

It looks like MoltenVk cannot find a descriptor set. I have narrowed it down to a `vkCmdBindDescriptorSets` command. 

I created a test scenario, in which I bind two sets: set 0 for global data and set 1 for the diffuse map sampler in the frag shader. I made it so that both set 0 and set 1 are created identically. Set 0 generally works in my engine. It has, among others, a descriptor for a texture in it, so I'm using this to sample from something in the frag shader. When I bind the global set a second time with the above `vkCmdBindDescriptorSets` instruction, it does not crash and I can sample from it in the frag shader with a test cube. The weird part is that when I use an identically created descriptor set, I get the seg fault.

**Comment by beaumanvienna** [comment]
I found the bug, I called `vkDestroyDescriptorSetLayout(device, m_DescriptorSetLayout, nullptr);` before `vkCmdBindDescriptorSets `.

During the initialization of the engine, I constructed the RAII wrapper for descriptor set layouts on the stack. The wrapper nuked the layout before the render loop even started.

The remaining question is, should we open a ticket about this behavior? It just crashed on `vkQueueSubmit` (see crash log above) with little clues as to what the issues is. There aren't many clues either in the Vulkan spec that the layout must be kept alive. This is just some food for thought, but I think MoltenVK could handle this user error more gracefully.

Anyways, for those who did read this here, thanks a lot!

**Comment by billhollings** [comment]
Transferring to issue #2356.
